core/src/main/java/org/bitcoinj/core/Locktime.java

```diff
@@ -0,0 +1,152 @@
+ * </p>
+ * <p>
+ * Instances of this class are immutable.
+ * </p>
+ */
+public class Locktime {
+    /**
+     * Wrap a raw value (as used in the Bitcoin protocol) into a lock time.
+     * @param rawValue raw value to be wrapped
+     * @return wrapped value
+     */
+    public static Locktime of(long rawValue) {
+        return new Locktime(rawValue);
+    }
+
+    /**
+     * Wrap a block height into a lock time.
+     * @param blockHeight block height to be wrapped
+     * @return wrapped block height
+     */
+    public static Locktime ofBlockHeight(int blockHeight) {
+        if (blockHeight >= THRESHOLD)
+            throw new IllegalArgumentException("block height too high: " + blockHeight);
+        return of(blockHeight);
+    }
+
+    /**
+     * Wrap a timestamp into a lock time.
+     * @param time timestamp to be wrapped
+     * @return wrapped timestamp
+     */
+    public static Locktime ofTime(Instant time) {
+        long secs = time.getEpochSecond();
+        if (secs < THRESHOLD)
+            throw new IllegalArgumentException("timestamp too low: " + secs);
+        return of(secs);
+    }
+
+    /**
+     * Construct an unset lock time.
+     * @return unset lock time
+     */
+    public static Locktime unset() {
+        return of(0);
+    }
+
+    private final long value;
+
+    // raw values below this threshold specify a block height, otherwise a timestamp in seconds since epoch
+    private static final int THRESHOLD = 500000000; // Tue Nov  5 00:53:20 1985 UTC
+
+    private Locktime(long rawValue) {
+        this.value = rawValue;
+    }
+
+    public long asRawValue() {
+        return value;
+    }
+
+    /**
+     * The lock time is considered to be set only if its raw value is greater than zero.
+     * @return true if lock time is set
+     */
+    public boolean isSet() {
+        return value > 0;
+    }
+
+    /**
+     * Determine if this lock time is specified as a block height.
+     * @return true if specified as a block height
+     */
+    public boolean isSpecifiedAsBlockHeight() {
+        return value < THRESHOLD;
+    }
+
+    /**
+     * Gets the lock time as a block height.
+     * @return lock time as a block height
+     * @throws IllegalStateException if the lock time is not specified as a block height
+     */
+    public int getAsBlockHeight() {
+        if (!isSpecifiedAsBlockHeight())
+            throw new IllegalStateException("lock time not specified as a block height");
+        return Math.toIntExact(value);
+    }
+
+    /**
+     * Determine if this lock time is specified as a timestamp in seconds since epoch.
+     * @return true if specified as a timestamp
+     */
+    public boolean isSpecifiedAsTime() {
+        return value >= THRESHOLD;
+    }
+
+    /**
+     * Gets the lock time as a timestamp.
+     * @return lock time as a timestamp
+     * @throws IllegalStateException if the lock time is not specified as a timestamp
+     */
+    public Instant getAsTime() {
```

Comment:
Make `Optional<>`. Consider shortening.

Reply:
Renamed to `timestamp()`. Note the "time" → "timestamp" change here, I think that's more congruent with the JavaDocs.

Reply:
I'll resolve this, as the remaining `Optional` change is contained in the to-do list I just posted.